src/haswell/avx2_convert_utf32_to_utf16.cpp

```diff
@@ -0,0 +1,51 @@
+std::pair<const char32_t*, char16_t*> avx2_convert_utf32_to_utf16(const char32_t* buf, size_t len, char16_t* utf16_output) {
+  const char32_t* end = buf + len;
+
+  const size_t safety_margin = 11; // to avoid overruns, see issue https://github.com/simdutf/simdutf/issues/92
+
+  while (buf + 8 + safety_margin <= end) {
+    __m256i in = _mm256_loadu_si256((__m256i*)buf);
+
+    const __m256i v_00000000 = _mm256_setzero_si256();
+    const __m256i v_ffff0000 = _mm256_set1_epi32((int32_t)0xffff0000);
+
+    // no bits set above 16th bit <=> can pack to UTF16 without surrogate pairs
+    const __m256i saturation_bytemask = _mm256_cmpeq_epi32(_mm256_and_si256(in, v_ffff0000), v_00000000);
+    const uint32_t saturation_bitmask = static_cast<uint32_t>(_mm256_movemask_epi8(saturation_bytemask));
+
+    if (saturation_bitmask == 0xffffffff) {
+      // validation
+      const __m256i v_f800 = _mm256_set1_epi32((int32_t)0xf800);
+      const __m256i v_d800 = _mm256_set1_epi32((int32_t)0xd800);
+      const __m256i forbidden_bytemask = _mm256_cmpeq_epi32(_mm256_and_si256(in, v_f800), v_d800);
+      const uint32_t forbidden_bitmask = static_cast<uint32_t>(_mm256_movemask_epi8(forbidden_bytemask));
+      if (forbidden_bitmask != 0) { return std::make_pair(nullptr, utf16_output); }
```

Comment:
You can save some work. By doing something like

```
forbidden_bytemask = _mm256_or_si256 (forbidden_bytemask, _mm256_and_si256(in, v_f800), v_d800)
```

and then checking forbidden_bytemask only at the very end. But it may not be worth the effort and maybe what you are doing is fine.


We can similarly check for the upper bound 0x10ffff in a similar (check manner).
```
runningmax = _mm256_max_epu32(_mm256_max_epu32(in,nextin), runningmax);
```

Then at the very end...
```
v10ffff = _mm256_set1_epi32(0x10ffff);
if(_mm256_movemask_epi8(_mm256_cmpeq_epi32(runningmax, v10ffff), v10ffff)) != 0xFFFFFFFF) {
// one word exceeds 0x10ffff and you must abort
}
```